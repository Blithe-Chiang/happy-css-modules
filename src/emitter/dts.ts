import { EOL } from 'os';
import { join, relative, basename } from 'path';
import camelcase from 'camelcase';
import { SourceNode, CodeWithSourceMap } from '../library/source-map/index.js';
import { Token } from '../loader.js';
import { LocalsConvention } from '../runner.js';
import { DistOptions, getRelativePath, isSubDirectoryFile, DtsFormatOptions } from './index.js';

/**
 * Get .d.ts file path.
 * @param filePath The path to the source file (i.e. `/dir/foo.css`). It is absolute.
 * @param distOptions The distribution option.
 * @returns The path to the .d.ts file. It is absolute.
 */
export function getDtsFilePath(filePath: string, distOptions: DistOptions | undefined): string {
  if (distOptions) {
    if (!isSubDirectoryFile(distOptions.rootDir, filePath))
      throw new Error(`The filePath(${filePath}) is not a subdirectory of rootDir(${distOptions.rootDir}).`);
    if (!isSubDirectoryFile(distOptions.rootDir, distOptions.outDir))
      throw new Error(`The outDir(${distOptions.outDir}) is not a subdirectory of rootDir(${distOptions.rootDir}).`);
    return join(distOptions.outDir, relative(distOptions.rootDir, filePath) + '.d.ts');
  } else {
    return filePath + '.d.ts';
  }
}

function dashesCamelCase(str: string): string {
  return str.replace(/-+(\w)/g, function (match, firstLetter) {
    return firstLetter.toUpperCase();
  });
}

function formatTokens(tokens: Token[], localsConvention: LocalsConvention): Token[] {
  const result: Token[] = [];
  for (const token of tokens) {
    if (localsConvention === 'camelCaseOnly') {
      result.push({ ...token, name: camelcase(token.name) });
    } else if (localsConvention === 'camelCase') {
      result.push(token);
      result.push({ ...token, name: camelcase(token.name) });
    } else if (localsConvention === 'dashesOnly') {
      result.push({ ...token, name: dashesCamelCase(token.name) });
    } else if (localsConvention === 'dashes') {
      result.push(token);
      result.push({ ...token, name: dashesCamelCase(token.name) });
    } else {
      result.push(token); // asIs
    }
  }
  return result;
}

function generateTokenDeclarations(
  sourceMapFilePath: string,
  tokens: Token[],
  dtsFormatOptions: DtsFormatOptions | undefined,
): typeof SourceNode[] {
  const formattedTokens = formatTokens(tokens, dtsFormatOptions?.localsConvention);
  const result: typeof SourceNode[] = [];

  for (const token of formattedTokens) {
    // Only one original position can be associated with one generated position.
    // This is due to the sourcemap specification. Therefore, we output multiple type definitions
    // with the same name and assign a separate original position to each.

    // NOTE: `--namedExport` does not support multiple jump destinations
    // TODO: Support multiple jump destinations with `--namedExport`
    for (const originalLocation of token.originalLocations) {
      if (dtsFormatOptions?.namedExport) {
        result.push(
          new SourceNode(null, null, null, [
            'export const ',
            new SourceNode(
              originalLocation.start.line ?? null,
              // The SourceNode's column is 0-based, but the originalLocation's column is 1-based.
              originalLocation.start.column - 1 ?? null,
              getRelativePath(sourceMapFilePath, originalLocation.filePath),
              `${token.name}`,
              token.name,
            ),
            ': string;',
          ]),
        );
      } else {
        result.push(
          new SourceNode(null, null, null, [
            'readonly ',
            new SourceNode(
              originalLocation.start.line ?? null,
              // The SourceNode's column is 0-based, but the originalLocation's column is 1-based.
              originalLocation.start.column - 1 ?? null,
              getRelativePath(sourceMapFilePath, originalLocation.filePath),
              `"${token.name}"`,
              token.name,
            ),
            ': string;',
          ]),
        );
      }
    }
  }
  return result;
}

export function generateDtsContentWithSourceMap(
  filePath: string,
  dtsFilePath: string,
  sourceMapFilePath: string,
  tokens: Token[],
  dtsFormatOptions: DtsFormatOptions | undefined,
): { dtsContent: CodeWithSourceMap['code']; sourceMap: CodeWithSourceMap['map'] } {
  const tokenDeclarations = generateTokenDeclarations(sourceMapFilePath, tokens, dtsFormatOptions);

  let sourceNode: typeof SourceNode;
  if (!tokenDeclarations || !tokenDeclarations.length) {
    sourceNode = new SourceNode(null, null, null, '');
  } else if (dtsFormatOptions?.namedExport) {
    sourceNode = new SourceNode(1, 0, getRelativePath(sourceMapFilePath, filePath), [
      'export const __esModule: true;' + EOL,
      ...tokenDeclarations.map((tokenDeclaration) => [tokenDeclaration, EOL]),
    ]);
  } else {
    sourceNode = new SourceNode(1, 0, getRelativePath(sourceMapFilePath, filePath), [
      'declare const styles: {' + EOL,
      ...tokenDeclarations.map((tokenDeclaration) => ['  ', tokenDeclaration, EOL]),
      '};' + EOL,
      'export = styles;' + EOL,
    ]);
  }
  const codeWithSourceMap = sourceNode.toStringWithSourceMap({
    // Since sourcemap and type definitions are in the same directory, they can be referenced by relative paths.
    file: basename(dtsFilePath),
    sourceRoot: '',
  });
  return {
    dtsContent: codeWithSourceMap.code,
    sourceMap: codeWithSourceMap.map,
  };
}
