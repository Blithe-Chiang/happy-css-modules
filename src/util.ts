/**
 * The SystemError type of Node.js.
 * @see https://nodejs.org/api/errors.html#class-systemerror
 */
export interface SystemError {
  code: string;
}

export function isSystemError(value: unknown): value is SystemError {
  return (
    isObject(value) &&
    hasProp(value, 'constructor') &&
    isObject(value.constructor) &&
    hasProp(value.constructor, 'name') &&
    value.constructor.name === 'Error' &&
    hasProp(value, 'code') &&
    typeof value.code === 'string'
  );
}

export function isObject(value: unknown): value is object {
  return (typeof value === 'object' || typeof value === 'function') && value !== null;
}

export function hasProp<T extends string>(obj: object, prop: T): obj is { [key in T]: unknown } {
  return prop in obj;
}

export function unique<T>(array: T[]): T[] {
  return [...new Set(array)];
}

export function uniqueBy<T, U>(arr: T[], fn: (el: T) => U): T[] {
  const result: T[] = [];
  const keys = new Set<U>();
  for (const el of arr) {
    const key = fn(el);
    if (!keys.has(key)) {
      keys.add(key);
      result.push(el);
    }
  }
  return result;
}
