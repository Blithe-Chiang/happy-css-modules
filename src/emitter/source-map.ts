import { EOL } from 'os';
import { getDtsFilePath } from './dts.js';
import { DistOptions, getRelativePath } from './index.js';

/**
 * Get .d.ts.map file path.
 * @param filePath The path to the source file (i.e. `foo.css`). It is absolute.
 * @param distOptions The distribution option.
 * @returns The path to the .d.ts.map file. It is absolute.
 */
export function getSourceMapFilePath(filePath: string, distOptions: DistOptions | undefined): string {
  return getDtsFilePath(filePath, distOptions) + '.map';
}

export function generateSourceMappingURLComment(dtsFilePath: string, sourceMapFilePath: string): string {
  return `//# sourceMappingURL=${getRelativePath(dtsFilePath, sourceMapFilePath)}` + EOL;
}
