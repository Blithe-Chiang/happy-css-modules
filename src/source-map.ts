import { SourceNode as OriginalSourceNode, CodeWithSourceMap } from 'source-map';

// The type definitions bundled in the source-map package are incorrect.
// Therefore, the type definitions are overwritten here.

type Chunk = string | StrictlyTypedSourceNode | Chunk[];

declare class StrictlyTypedSourceNode extends OriginalSourceNode {
  constructor(line: number | null, column: number | null, source: string | null);
  constructor(line: number | null, column: number | null, source: string | null, chunk?: Chunk, name?: string);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const SourceNode: StrictlyTypedSourceNode = OriginalSourceNode as any;

export { SourceNode, CodeWithSourceMap };
