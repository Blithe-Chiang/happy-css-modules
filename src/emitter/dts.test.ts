import dedent from 'dedent';
import { SourceMapConsumer } from 'source-map';
import { Loader } from '../loader/index.js';
import { getFixturePath, createFixtures } from '../test/util.js';
import { generateDtsContentWithSourceMap, getDtsFilePath } from './dts.js';
import { type DtsFormatOptions } from './index.js';

const loader = new Loader();

test('getDtsFilePath', () => {
  expect(getDtsFilePath('/app/src/dir/1.css', undefined)).toBe('/app/src/dir/1.css.d.ts');
  expect(getDtsFilePath('/app/src/dir/1.css', { rootDir: '/app', outDir: '/app/dist' })).toBe(
    '/app/dist/src/dir/1.css.d.ts',
  );
  expect(() => getDtsFilePath('/tmp/src/dir/1.css', { rootDir: '/app', outDir: '/app/dist' })).toThrow();
  expect(() => getDtsFilePath('/app/src/dir/1.css', { rootDir: '/app', outDir: '/tmp/dist' })).toThrow();
});

describe('generateDtsContentWithSourceMap', () => {
  const filePath = getFixturePath('/test/1.css');
  const dtsFilePath = getFixturePath('/test/1.css.d.ts');
  const sourceMapFilePath = getFixturePath('/test/1.css.map');
  const dtsFormatOptions: DtsFormatOptions = {
    localsConvention: undefined,
  };
  test('generate dts content with source map', async () => {
    createFixtures({
      '/test/1.css': dedent`
      @import './2.css';
      .a {}
      .b {}
      .b {}
      `,
      '/test/2.css': dedent`
      @import './3.css';
      .c {}
      `,
      '/test/3.css': dedent`
      .d {}
      `,
    });
    const result = await loader.load(filePath);
    const { dtsContent, sourceMap } = generateDtsContentWithSourceMap(
      filePath,
      dtsFilePath,
      sourceMapFilePath,
      result.tokens,
      dtsFormatOptions,
    );
    expect(dtsContent).toMatchInlineSnapshot(`
      "declare const styles: {
        readonly "d": string;
        readonly "c": string;
        readonly "a": string;
        readonly "b": string;
        readonly "b": string;
      };
      export default styles;
      "
    `);
    const smc = await new SourceMapConsumer(sourceMap.toJSON());
    expect(smc.originalPositionFor({ line: 2, column: 11 })).toMatchInlineSnapshot(`
      {
        "column": 0,
        "line": 1,
        "name": "d",
        "source": "3.css",
      }
    `);
    expect(smc.originalPositionFor({ line: 3, column: 11 })).toMatchInlineSnapshot(`
      {
        "column": 0,
        "line": 2,
        "name": "c",
        "source": "2.css",
      }
    `);
    expect(smc.originalPositionFor({ line: 4, column: 11 })).toMatchInlineSnapshot(`
      {
        "column": 0,
        "line": 2,
        "name": "a",
        "source": "1.css",
      }
    `);
    expect(smc.originalPositionFor({ line: 5, column: 11 })).toMatchInlineSnapshot(`
      {
        "column": 0,
        "line": 3,
        "name": "b",
        "source": "1.css",
      }
    `);
    expect(smc.originalPositionFor({ line: 6, column: 11 })).toMatchInlineSnapshot(`
      {
        "column": 0,
        "line": 4,
        "name": "b",
        "source": "1.css",
      }
    `);
  });
  describe('format case', () => {
    async function getResult(filePath: string) {
      createFixtures({
        '/test/1.css': dedent`
        .foo-bar {}
        .foo_bar {}
        `,
      });
      return await loader.load(filePath);
    }
    test('undefined', async () => {
      const result = await getResult(filePath);
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, result.tokens, {
        ...dtsFormatOptions,
        localsConvention: undefined,
      });
      expect(dtsContent).toMatchInlineSnapshot(`
        "declare const styles: {
          readonly "foo-bar": string;
          readonly "foo_bar": string;
        };
        export default styles;
        "
      `);
    });
    test('camelCaseOnly', async () => {
      const result = await getResult(filePath);
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, result.tokens, {
        ...dtsFormatOptions,
        localsConvention: 'camelCaseOnly',
      });
      expect(dtsContent).toMatchInlineSnapshot(`
        "declare const styles: {
          readonly "fooBar": string;
          readonly "fooBar": string;
        };
        export default styles;
        "
      `);
    });
    test('camelCase', async () => {
      const result = await getResult(filePath);
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, result.tokens, {
        ...dtsFormatOptions,
        localsConvention: 'camelCase',
      });
      expect(dtsContent).toMatchInlineSnapshot(`
        "declare const styles: {
          readonly "foo-bar": string;
          readonly "fooBar": string;
          readonly "foo_bar": string;
          readonly "fooBar": string;
        };
        export default styles;
        "
      `);
    });
    test('dashesOnly', async () => {
      const result = await getResult(filePath);
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, result.tokens, {
        ...dtsFormatOptions,
        localsConvention: 'dashesOnly',
      });
      expect(dtsContent).toMatchInlineSnapshot(`
        "declare const styles: {
          readonly "fooBar": string;
          readonly "foo_bar": string;
        };
        export default styles;
        "
      `);
    });
    test('dashes', async () => {
      const result = await getResult(filePath);
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, result.tokens, {
        ...dtsFormatOptions,
        localsConvention: 'dashes',
      });
      expect(dtsContent).toMatchInlineSnapshot(`
        "declare const styles: {
          readonly "foo-bar": string;
          readonly "fooBar": string;
          readonly "foo_bar": string;
          readonly "foo_bar": string;
        };
        export default styles;
        "
      `);
    });
  });
  test('emit other directory', async () => {
    createFixtures({
      '/test/1.css': `.a {}`,
    });
    const result = await loader.load(filePath);
    const { dtsContent, sourceMap } = generateDtsContentWithSourceMap(
      getFixturePath('/test/src/1.css'),
      getFixturePath('/test/dist/1.css.d.ts'),
      getFixturePath('/test/dist/1.css.d.ts.map'),
      result.tokens,
      dtsFormatOptions,
    );
    expect(dtsContent).toMatchInlineSnapshot(`
      "declare const styles: {
        readonly "a": string;
      };
      export default styles;
      "
    `);
    const smc = await new SourceMapConsumer(sourceMap.toJSON());
    // FIXME: `source` should be `../src/1.css`
    expect(smc.originalPositionFor({ line: 2, column: 11 })).toMatchInlineSnapshot(`
      {
        "column": 0,
        "line": 1,
        "name": "a",
        "source": "../1.css",
      }
    `);
  });
});
