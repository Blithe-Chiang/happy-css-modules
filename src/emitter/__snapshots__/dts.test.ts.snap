// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDtsContentWithSourceMap change export type by namedExport 2`] = `
{
  "version": 3,
  "sources": ["1.css"],
  "names": ["foo", "bar"],
  "mappings": "AAAA;AAAA,E,SAAAA,K,SAAA;AAAA,E,SACAC,K,SADA;AAAA;AAAA",
  "file": "1.css.d.ts",
  "sourceRoot": ""
}
`;

exports[`generateDtsContentWithSourceMap change export type by namedExport 4`] = `
{
  "version": 3,
  "sources": ["1.css"],
  "names": ["foo", "bar"],
  "mappings": "AAAA;aAAAA,G,SAAA;aACAC,G,SADA",
  "file": "1.css.d.ts",
  "sourceRoot": ""
}
`;

exports[`generateDtsContentWithSourceMap emit other directory 2`] = `
{
  "version": 3,
  "sources": ["../src/1.css"],
  "names": ["foo", "bar"],
  "mappings": "AAAA;AAAA,E,SAAAA,K,SAAA;AAAA,E,SACAC,K,SADA;AAAA;AAAA",
  "file": "1.css.d.ts",
  "sourceRoot": ""
}
`;

exports[`generateDtsContentWithSourceMap format case camelCase 1`] = `
"declare const styles: {
  readonly \\"foo-bar\\": string;
  readonly \\"fooBar\\": string;
  readonly \\"foo_bar\\": string;
  readonly \\"fooBar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap format case camelCaseOnly 1`] = `
"declare const styles: {
  readonly \\"fooBar\\": string;
  readonly \\"fooBar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap format case dashes 1`] = `
"declare const styles: {
  readonly \\"foo-bar\\": string;
  readonly \\"fooBar\\": string;
  readonly \\"foo_bar\\": string;
  readonly \\"foo_bar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap format case dashesOnly 1`] = `
"declare const styles: {
  readonly \\"fooBar\\": string;
  readonly \\"foo_bar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap format case undefined 1`] = `
"declare const styles: {
  readonly \\"foo-bar\\": string;
  readonly \\"foo_bar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap generate dts content with source map 1`] = `
"declare const styles: {
  readonly \\"foo\\": string;
  readonly \\"bar\\": string;
  readonly \\"baz\\": string;
  readonly \\"baz\\": string;
  readonly \\"qux\\": string;
  readonly \\"quux\\": string;
  readonly \\"quux\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap generate dts content with source map 2`] = `
{
  "version": 3,
  "sources": ["1.css", "2.css"],
  "names": ["foo", "bar", "baz", "qux", "quux"],
  "mappings": "AAAA;AAAA,E,SAAAA,K,SAAA;AAAA,E,SACAC,K,SADA;AAAA,E,SAEAC,K,SAFA;AAAA,E,SAGAA,K,SAHA;AAAA,E,SCIAC,K,SDJA;AAAA,E,SCKAC,M,SDLA;AAAA,E,SCMAA,M,SDNA;AAAA;AAAA",
  "file": "1.css.d.ts",
  "sourceRoot": ""
}
`;
